utility/servicer/module.go

```diff
@@ -138,19 +235,86 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
 	return nil
 }
 
-// TODO: implement
 // validateApplication makes sure the application has not received more relays than allocated in the current session.
-func (s servicer) validateApplication(meta *coreTypes.RelayMeta, session *coreTypes.Session) error {
-	/*
-		// if maxRelaysPerSession, overServiced := calculateAppSessionTokens(); overServiced {
-			return fmt.Errorf("application %s has exceeded its allocated relays %d for the session %d", meta.ApplicationPublicKey, maxRelaysPerSession)
-		}
-	*/
-	return nil
+func (s *servicer) validateApplication(session *coreTypes.Session) error {
+	// IMPROVE: use a function to get current height from the current session
+	servicerAppSessionTokens, err := s.calculateServicerAppSessionTokens(session)
+	if err != nil {
+		return fmt.Errorf("Error calculating servicer tokens for application: %w", err)
+	}
+
+	localCtx, err := s.GetBus().GetPersistenceModule().NewLocalContext()
+	if err != nil {
+		return fmt.Errorf("Error getting local persistence context: application %s session number %d: %w", session.Application.PublicKey, session.SessionNumber, err)
+	}
+
+	usedAppSessionTokens, err := localCtx.GetSessionTokensUsed(session)
+	if err != nil {
+		return fmt.Errorf("Error getting servicer token usage: application %s session number %d: %w", session.Application.PublicKey, session.SessionNumber, err)
+	}
+
+	if usedAppSessionTokens == nil || usedAppSessionTokens.Cmp(servicerAppSessionTokens) < 0 {
+		return nil
+	}
+
+	return fmt.Errorf("application %s has exceeded its allocated relays %s for session %d",
+		session.Application.PublicKey,
+		servicerAppSessionTokens,
+		session.SessionNumber)
+}
+
+func (s *servicer) cachedAppTokens(session *coreTypes.Session) *sessionTokens {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+
+	return s.totalTokens[session.Application.PublicKey]
+}
+
+// calculateServicerAppSessionTokens return the number of tokens the servicer has remaining for the Application in the session provided.
+//
+//	If nothing is cached, the maximum number of session tokens is computed.
+func (s *servicer) calculateServicerAppSessionTokens(session *coreTypes.Session) (*big.Int, error) {
+	tokens := s.cachedAppTokens(session)
+	if tokens != nil && tokens.originalCountAvailable != nil && tokens.sessionNumber == session.SessionNumber {
+		return big.NewInt(1).Set(tokens.originalCountAvailable), nil
```

Comment:
This returns the original count available but the comment says "number of tokens remaining".

Either the comment or business logic is off. Might also be worth updating the function name to what this is intended to do.

Reply:
Thank you, fixed both the comment and the function name.